src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -34,6 +44,82 @@ import           Database.LSMTree.Internal.RunNumber
 import           System.FS.API (HasFS)
 import           System.FS.BlockIO.API (HasBlockIO)
 
+{-------------------------------------------------------------------------------
+  Versioning
+-------------------------------------------------------------------------------}
+
+-- | The version of a snapshot.
+--
+-- When encoding snapshot metadata, 'currentSnapshotVersion' is included in the
+-- file. When snapshot metadata is decoded, then the decoded version is compared
+-- against 'currentSnapshotVersion'. If the versions do not match, then decoding
+-- may fail, but the version information should be sufficient to check backwards
+-- compatibility. That is, if 'currentSnapshotVersions' is greater than the
+-- decoded version, then decoding might succeed. Forwards compatibility is never
+-- guaranteed: decoding a newer snapshot with an older version of the library
+-- should always fail.
+--
+-- A version @x.y@ has two components: a major version number @x@ and a minor
+-- version number @y@. @x.y@ is guaranteed to be backwards compatible with
+-- @x'.y'@ as long as @x == x'@ and @y' <= y@. If @x > x'@, then backwards
+-- compatibility *may* be provided, but is not guaranteed.
+--
+-- The version number determines the format of the snapshot metadata file, but
+-- it also doubles as versioning information for the entire snapshot itself. For
+-- example, if a breaking change is made to the merge scheduling algorithm that
+-- would lead to errors when loading an older snapshot, then the major version
+-- should be increased as well.
```

Comment:
So if I understand correctly:
- the major version is necessary because we want to be able to make breaking changes
- the minor version is useful because we can make changes that are backwards compatible, but not forward compatible, plus it might help with debugging things to know exactly which version something is.

Is that correct?

Reply:
Yup, that's about it. Should I clarify more in the comment?

Reply:
That would be great.

Reply:
Updated. Let me know what you think

Reply:
:+1: